File: flutter_sdk/lib/src/common/result.dart
// Copyright (c) Yalochat, Inc. All rights reserved.

import 'package:equatable/equatable.dart';

sealed class Result<T> {
  const Result();

  const factory Result.ok(T value) = Ok._;

  const factory Result.error(Exception error) = Error._;
}

final class Ok<T> extends Result<T> with EquatableMixin {
  final T result;
  const Ok._(this.result);

  @override
  List<Object?> get props => [result];
}

final class Error<T> extends Result<T> with EquatableMixin {
  final Exception error;
  const Error._(this.error);

  @override
  List<Object?> get props => [error];
}

File: typescript-sdk/packages/core/src/common/result.ts
// Copyright (c) Yalochat, Inc. All rights reserved.

export type Result<T> = { ok: true; value: T } | { ok: false; error: Error };

export const ok = <T>(value: T): Result<T> => ({ ok: true, value });
export const err = (error: Error): Result<never> => ({ ok: false, error });
